lib/pure/sugar.nim

```diff
@@ -198,3 +198,65 @@ macro dump*(x: typed): untyped =
   let r = quote do:
     debugEcho `s`, " = ", `x`
   return r
+
+# TODO: consider exporting this in macros.nim
+proc replaceNodes(ast: NimNode): NimNode =
+  # Replace NimIdent and NimSym by a fresh ident node
+  # see also https://github.com/nim-lang/Nim/pull/8531#issuecomment-410436458
+  proc inspect(node: NimNode): NimNode =
+    case node.kind:
+    of nnkIdent, nnkSym:
+      return ident($node)
+    of nnkEmpty, nnkLiterals:
+      return node
+    else:
+      var rTree = node.kind.newTree()
+      for child in node:
+        rTree.add inspect(child)
+      return rTree
+  result = inspect(ast)
+
+macro distinctBase*(T: typedesc, recursive: static[bool] = false): untyped =
+  ## reverses ``type T = distinct A``
+  runnableExamples:
+    type T = distinct int
+    doAssert distinctBase(T) is int
+    doAssert: not compiles(distinctBase(int))
+    type T2 = distinct T
+    doAssert distinctBase(T2, recursive = false) is T
+    doAssert distinctBase(int, recursive = true) is int
+    doAssert distinctBase(T2, recursive = true) is int
+
+  let typeNode = getTypeImpl(T)
+  expectKind(typeNode, nnkBracketExpr)
+  if $typeNode[0] != "typeDesc":
+    error "expected typeDesc, got " & $typeNode[0]
+  var typeSym = typeNode[1]
+  if not recursive:
+    let impl = getTypeImpl(typeSym)
+    if $impl.typeKind != "ntyDistinct":
+      error "type is not distinct"
+    typeSym = getTypeInst(impl[0])
+  else:
+    while true:
+      let impl = getTypeImpl(typeSym)
+      if $impl.typeKind != "ntyDistinct":
+        typeSym = impl
+        break
+      typeSym=getTypeInst(impl[0])
+  typeSym.replaceNodes
+
+proc distinctBase*[T](a: T, recursive: static[bool] = false): auto =
```

Comment:
@araq indeed, this feature was suggested by @mratsim and I agree with him that it's useful. 

Changed to `Here I think func {.inline.}` (not template, according to rule of preferring procs/funcs whenever possible)

> basically a no-op in the C backend

I would hope with high enough optimization level the `{.inline.}` would not make a difference though

> You cannot even write down its return type explicitly.

that's because of https://github.com/nim-lang/Nim/issues/8545

so I kept `auto`